keps/sig-storage/4876-mutable-csinode-allocatable/README.md

```diff
@@ -0,0 +1,845 @@
+
+- Allow users to opt in to this feature at a per-CSI driver granularity by configuring the `CSIDriver` object. Specifically, administrators will be able to fine-tune the interval update value via the `NodeAllocatableUpdatePeriodSeconds` attribute in the `CSIDriver` object as per their specific requirement.
+
+- A reactive update mechanism is implemented to immediately update the `CSINode.Spec.Drivers[*].Allocatable.Count` value if a pod fails to enter a running state due volume attachment failures as a result of insufficient capacity. This ensures that even if a race occurs, Kubernetes quickly corrects itself and prevents further scheduling errors.
+
+## Design Details
+
+### Feature Gate
+
+A new feature gate - `MutableCSINodeAllocatableCount` - will be introduced to control the functionality implemented by this KEP. When the feature gate is disabled, the `CSINode` object will remain immutable, maintaining the current behavior.
+
+### API Changes
+
+#### CSINode
+
+The `CSINode.Spec.Drivers[*].Allocatable.Count` field will be made mutable. No changes to the object structs are needed, only the validation logic needs to be revised. For reference, these are the API fields this KEP proposes to make mutable.
+
+```golang
+// CSINodeDriver holds information about the specification of one CSI driver installed
+type CSINodeDriver struct {
+    ...
+    // allocatable represents the volume resources of a node that are available for sc
+    // +optional
+    Allocatable *VolumeNodeResources
+}
+```
+
+```golang
+// VolumeNodeResources is a set of resource limits for scheduling of volumes.
+type VolumeNodeResources struct {
+    // Maximum number of unique volumes managed by the CSI driver that can be used on
+    // A volume that is both attached and mounted on a node is considered to be used o
+    // The same rule applies for a unique volume that is shared among multiple pods on
+    // If this field is not specified, then the supported number of volumes on this no
+    // +optional
+    Count *int32
+}
+```
+
+#### CSIDriver
+
+A new field, `NodeAllocatableUpdatePeriodSeconds`, will be added to the `CSIDriverSpec` struct. This field allows a CSI driver to specify the interval at which the Kubelet should periodically query a driver's `NodeGetInfo` RPC endpoint to update the `CSINode` object. If this field is not set, updates will only occur in response to volume attachment failures as a result of no capacity.
+
+```golang
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    ...
+    // NodeAllocatableUpdatePeriodSeconds specifies the interval between periodic updates of
+    // the CSINode allocatable capacity for this driver. If not set, periodic updates
+    // are disabled, and updates occur only upon detecting capacity-related failures.
+    // The minimum allowed value for this field is 10 seconds.
+    // +optional
+    NodeAllocatableUpdatePeriodSeconds *metav1.Duration
```

Comment:
Would it be better to use a stream for CSIDriver to report back changes? Same way as we do Device Health in Device Plugin?

Reply:
None of the CSI RPCs support streaming IIRC.